File: src/hyperliquid_python/functions/cleaner.py
from repositories.dynamo_repository import list_all_open_orders, update_order
from helpers.hyperliquid_helpers import get_env_config, create_clients

def handler(event, context):
    print("Reconciliation job started...")
    
    try:
        private_key, user_address, is_testnet = get_env_config()
        _, info_client = create_clients(private_key, is_testnet)
        
        db_open_trades = list_all_open_orders()
        if not db_open_trades:
            print("No open trades found in database. Finishing job.")
            return

        hl_open_orders = info_client.open_orders(user_address)
        hl_open_oids = set(str(order.get('oid')) for order in hl_open_orders)
        
        reconciled_count = 0
        
        for trade in db_open_trades:
            entry_oid_exists = str(trade.oid) in hl_open_oids if trade.oid else False
            stop_loss_oid_exists = str(trade.stopLossOid) in hl_open_oids if trade.stopLossOid else False
            
            if not entry_oid_exists and not stop_loss_oid_exists:
                print(f"Reconciling orphaned trade: Symbol={trade.symbol}, Strategy={trade.strategy}, ID={trade.id}")
                
                update_order(trade.id, status='closed')
                reconciled_count += 1
                
        print(f"Reconciliation job finished. Total trades checked: {len(db_open_trades)}, Reconciled: {reconciled_count}")
        
    except Exception as e:
        print(f"Error during trade reconciliation: {e}")
        raise e

File: src/hyperLiquid_ts_deprecated/functions/cleaner.ts
import { listAllOpenOrders, updateOrder } from "../repositories/dynamo.repository";
import { getEnvConfig, createClients } from "../helpers/hyperliquid.helpers";

/**
 * EventBridge-triggered Lambda to periodically reconcile open trades.
 * Runs every 6 hours to find trades closed outside of our webhook.
 */
export const handler = async (event: any, context: any): Promise<void> => {
    console.log("Reconciliation job started...");

    try {
        const { privateKey, userAddress, isTestnet } = getEnvConfig();
        const { infoClient } = createClients(privateKey, isTestnet);

        // Fetch all trades marked as 'open' in our database
        const dbOpenTrades = await listAllOpenOrders();
        if (dbOpenTrades.length === 0) {
            console.log("No open trades found in database. Finishing job.");
            return;
        }

        // Fetch all currently open orders from HyperLiquid API
        const hlOpenOrders = await infoClient.openOrders({ user: userAddress as `0x${string}` });

        // Use a Set for O(1) order ID lookup
        const hlOpenOids = new Set(hlOpenOrders.map(o => o.oid.toString()));

        let reconciledCount = 0;

        for (const trade of dbOpenTrades) {
            const entryOidExists = trade.oid ? hlOpenOids.has(trade.oid.toString()) : false;
            const stopLossOidExists = trade.stopLossOid ? hlOpenOids.has(trade.stopLossOid.toString()) : false;

            // If neither the main order nor the stop loss are active, the trade is finished
            if (!entryOidExists && !stopLossOidExists) {
                console.log(`Reconciling orphaned trade: Symbol=${trade.symbol}, Strategy=${trade.strategy}, ID=${trade.id}`);

                await updateOrder(trade.id, {
                    status: 'closed'
                });

                reconciledCount++;
            }
        }

        console.log(`Reconciliation job finished. Total trades checked: ${dbOpenTrades.length}, Reconciled: ${reconciledCount}`);

    } catch (error) {
        console.error("Error during trade reconciliation:", error instanceof Error ? error.message : error);
        throw error;
    }
}
